### PR TITLE
[PoC] Explore external event emitter contract

### PR DESCRIPTION
**Pull request by zerotucks**
SAMPLE TAKEN
Whilst running open aave test with steth:eth tokens.

DIFFERENCE
* Additional 120,000 Gas on Open AAVE Position. Roughly 120K ~ 150K per major operation.
* ~4.5% increase in gas cost

GAS **BEFORE** EVENT EMITTER
Screenshot 2022-11-24 at 13 56 07

GAS **WITH PROXY ADDR SET IN OP-EXEC** ON EVENT EMITTER
Screenshot 2022-11-28 at 11 19 23

### CODE REVIEW

File: contracts/actions/common/PullToken.sol [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
+import { EVENT_EMITTER, PULL_TOKEN_ACTION } from "../../core/constants/Common.sol";
 import "../../core/types/Common.sol";
+import { IEventEmitter } from "../../interfaces/common/IEventEmitter.sol";
 
 /**
  * @title PullToken Action contract
  * @notice Pulls token from a target address to the current calling context
  */
-contract PullToken is Executable {
+contract PullToken is Executable, UseStore {
```
Comment (peculiarity) [inline contracts/actions/common/PullToken.sol]: You don't have to implement `UseStore`.
Having `constructor (address _registry)` is just fine.

File: contracts/actions/common/ReturnFunds.sol [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
+import { EVENT_EMITTER, RETURN_FUNDS_ACTION, ETH } from "../../core/constants/Common.sol";
 import { DSProxy } from "../../libs/DS/DSProxy.sol";
+import { IEventEmitter } from "../../interfaces/common/IEventEmitter.sol";
 
 /**
  * @title ReturnFunds Action contract
  * @notice Returns funds sitting on a user's proxy to a user's EOA
  */
-contract ReturnFunds is Executable {
+contract ReturnFunds is Executable, UseStore {
```
Comment (peculiarity) [inline contracts/actions/common/ReturnFunds.sol]: Same..

File: contracts/actions/common/SendToken.sol [hunk 3]
```diff
@@ -1,17 +1,21 @@
...
-import { SEND_TOKEN_ACTION } from "../../core/constants/Common.sol";
+import { EVENT_EMITTER, SEND_TOKEN_ACTION } from "../../core/constants/Common.sol";
+import { IEventEmitter } from "../../interfaces/common/IEventEmitter.sol";
 
 /**
  * @title SendToken Action contract
  * @notice Transfer token from the calling contract to the destination address
  */
-contract SendToken is Executable {
+contract SendToken is Executable, UseStore {
```
Comment (peculiarity) [inline contracts/actions/common/SendToken.sol]: same

File: contracts/actions/common/SetApproval.sol [hunk 4]
```diff
@@ -38,7 +39,12 @@ contract SetApproval is Executable, UseStore {
 
     IERC20(approval.asset).safeApprove(approval.delegate, actualApprovalAmount);
 
-    emit Action(SET_APPROVAL_ACTION, bytes(abi.encode(actualApprovalAmount)));
+    IEventEmitter eventEmitter = IEventEmitter(registry.getRegisteredService(EVENT_EMITTER));
+    eventEmitter.emitActionEvent(
+      SET_APPROVAL_ACTION,
+      msg.sender,
```
Comment (peculiarity) [inline contracts/actions/common/SetApproval.sol]: Why `msg.sender` hre and not `address(this)` ?

File: contracts/core/EventEmitter.sol [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  function emitActionEvent(
+    string memory actionName,
+    address proxyAddress,
+    bytes calldata encodedReturnValues
+  ) external {
+    emit Action(actionName, proxyAddress, encodedReturnValues);
+  }
+
+  function emitOperationEvent(
```
Comment (peculiarity) [inline contracts/core/EventEmitter.sol:59]: Not sure about that. OperationRegistry is not called via Proxy.  We can track the opregistry directly.

File: contracts/core/EventEmitter.sol [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * proxy contract.
+ */
+contract EventEmitter {
+  /**
+   * @dev Emitted once an Action has completed execution
+   * @param name The Action name
+   * @param proxyAddress The proxy address of the user executing the t/x
+   * @param returned The bytes value returned by the Action
+   **/
+  event Action(string name, address proxyAddress, bytes returned);
```
Comment (peculiarity) [inline contracts/core/EventEmitter.sol]: `indexed` the name at least. Not sure about `proxyAddress`, I have no idea what's the purpose behind that.

File: contracts/actions/common/ReturnFunds.sol (hunk 2 above)
Comment (zerotucks) [reply contracts/actions/common/ReturnFunds.sol]: Mistake

File: contracts/actions/common/SetApproval.sol (hunk 4 above)
Comment (zerotucks) [reply contracts/actions/common/SetApproval.sol]: Mistake

File: contracts/core/EventEmitter.sol (hunk 6 above)
Comment (zerotucks) [reply contracts/core/EventEmitter.sol]: ProxyAddress is a requirement from Damian so it's easy to map back events to a given user.

File: contracts/core/EventEmitter.sol (hunk 5 above)
Comment (zerotucks) [reply contracts/core/EventEmitter.sol:59]: OpRegistry emits events when an Op is added but not when fetched/executed - that's done within OpExecutor

Comment (peculiarity) [reply contracts/core/EventEmitter.sol:59]: So this is something new. Any reasoning behind the need of that?

Comment (zerotucks) [reply contracts/core/EventEmitter.sol:59]: It's just the same functionality that it's in the OperationExecutor moved to this contract? Not sure I follow

**Pull request closed without merging**